src/services/cache-service.ts

```diff
@@ -2,63 +2,109 @@
  * Caches the response of any promise value of a loading function and prevents duplicate calls
  * @author Mike Reiche <mike.reiche@t-systems.com>
  */
+import {TimeComponents, toMilliseconds} from "../utils/time";
+
 export class CacheService {
-    private _cacheContainer = {};
-    private _defaultCacheTtlSeconds = 10;
+    private _cacheContainer:{[key: string]: [number, Promise<any>]} = {};
+    private _defaultCacheTtl:TimeComponents = {seconds: 10};
 
+    /**
+     *
+     * @param key The key for cached value
+     * @param loadingFunction The loading function returning a promise of the value.
+     * @param ttl The cache timeout for the response.
+     */
     getForKeyWithLoadingFunction <T> (
         key:string,
         loadingFunction:() => Promise<T>,
-        cacheTtlSeconds = this._defaultCacheTtlSeconds
+        ttl:number|TimeComponents = this._defaultCacheTtl
     ):Promise<T> {
-        let cacheEntry = this._cacheContainer[key];
-        const now = (Date.now()/1000);
-        if (cacheEntry !== undefined) {
-            if (cacheEntry[0] >= now - cacheTtlSeconds) {
-                return cacheEntry[1];
-            }
+        const cacheEntry = this._cacheContainer[key];
+        const now = Date.now();
+
+        if (cacheEntry !== undefined && cacheEntry[0] >= now) {
```

Comment:
My view on adding a "Pending" state or something to each cache entry, as suggested in #42 :
This test fails even though the service should prevent duplicate calls.
```ts
  test("Loading function is not called again if its response is pending", async () => {
    const cacheService = new CacheService()
    const key = "someKey";

    const ttl = {ms: 10}
    const loadingTime = {ms: 30}

    const mockFunction = jest.fn();

    const aLoadingFunction = async () => {
      mockFunction()
      await sleep(toMilliseconds(loadingTime));
      return Promise.resolve();
    }

    await cacheService.getForKeyWithLoadingFunction(key, aLoadingFunction, ttl)
    
    // The second call should not go through if the first one hasn't returned yet
    await cacheService.getForKeyWithLoadingFunction(key, aLoadingFunction, ttl)

    expect(mockFunction).toBeCalledTimes(1)
  })
```

Reply:
You wait for 30ms for the response of a cache entry, which has a ttl of 10ms, so its actually already outdated for 20ms when you get the response.
Removing the `await` synchronisation makes more sense here.

```typescript
test("Loading function is not called again if its response is pending", async () => {
    const cacheService = new CacheService()
    const key = "someKey";

    const ttl = {ms: 10}
    const loadingTime = {ms: 30}

    const mockFunction = jest.fn();

    const aLoadingFunction = async () => {
      mockFunction()
      await sleep(toMilliseconds(loadingTime));
      return Promise.resolve();
    }

    cacheService.getForKeyWithLoadingFunction(key, aLoadingFunction, ttl)

    // The second call should not go through if the first one hasn't returned yet
    await cacheService.getForKeyWithLoadingFunction(key, aLoadingFunction, ttl)

    expect(mockFunction).toBeCalledTimes(1)
  })
  ```

Reply:
Alright, I agree, when we keep the definition of "valid until" from above, this test does not map to a use case. 
Ignoring it is fine by me.